VirtoCommerce.PricingModule.Data/Services/PricingServiceImpl.cs

```diff
@@ -206,20 +228,59 @@ public virtual IEnumerable<coreModel.Price> EvaluateProductPrices(coreModel.Pric
                     }
                 }
             }
-                    
+
             return retVal;
         }
-            
+
+        public virtual IEnumerable<coreModel.PriceCalendarChange> GetCalendarChanges(DateTime? lastEvaluationTimestamp, DateTime? evaluationTimestamp, int? skip, int? take)
+        {
+            if (!AllowTimeFilters)
+            {
+                yield break;
+            }
+
+            using (var repository = _repositoryFactory())
+            {
+                repository.DisableChangesTracking();
+
+                // Increase command timeout to allow lengthy queries.
+                repository.UnitOfWork.CommandTimeout = TimeSpan.MaxValue;
+
+                lastEvaluationTimestamp = lastEvaluationTimestamp ?? DateTime.MinValue;
+
+                var query = repository.Prices
+                    .Where(x => (x.EndDate < evaluationTimestamp && x.EndDate > lastEvaluationTimestamp)
+                                || (x.StartDate <= evaluationTimestamp && x.StartDate > lastEvaluationTimestamp))
+                    .OrderBy(x => x.ProductId) as IQueryable<dataModel.PriceEntity>;
+
+                if (skip != null)
+                    query = query.Skip(skip.Value);
+                if (take != null)
+                    query = query.Take(take.Value);
+
+                var groupedQuery = query
+                 .Select(x => x.ProductId)
+                 .GroupBy(x => x);
+
+                foreach (var calendarChange in groupedQuery.AsNoTracking())
+                {
+                    yield return new coreModel.PriceCalendarChange
```

Comment:
I see what you mean, but when there are large sets of prices, doing paged queries iteratively will put a lot of load on the SQL Server. So all in all, this seems to be better.
wdyt?

Reply:
In any case, there will be only one database  request per each method call. Because skip and take parameters are passed as arguments to this method doesn't contains any other pagination logic

Reply:
But if we would not use yielding, we would need to use pagination, because otherwise we could have memory issues.